test/distributed/_tensor/test_attention.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+# Owner(s): ["oncall: distributed"]
+
+import torch
+from torch.distributed._tensor import DeviceMesh, DTensor, Replicate, Shard
+from torch.distributed._tensor.debug import CommDebugMode
+from torch.distributed._tensor.experimental.attention import attention_parallel
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    run_tests,
+)
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    NUM_DEVICES,
+    with_comms,
+)
+
+
+c10d_functional = torch.ops.c10d_functional
+
+
+class RingAttentionTest(DTensorTestBase):
+    @with_comms
+    def test_ring_attention(self):
+        device_mesh = DeviceMesh(
+            self.device_type,
+            torch.arange(0, NUM_DEVICES),
+        )
+        dtype = torch.bfloat16
+        bs = 8
+        query_tokens = 8
+        context_tokens = 16
+        dim = 32
+        nheads = 8
+        query = torch.rand(
+            (bs, nheads, query_tokens, dim),
+            device=self.device_type,
+            dtype=dtype,
+            requires_grad=True,
+        )
+        original_query = query
+        key = torch.rand(
+            (bs, nheads, context_tokens, dim), device=self.device_type, dtype=dtype
+        )
+        value = torch.rand(
+            (bs, nheads, context_tokens, dim), device=self.device_type, dtype=dtype
+        )
+
+        query_placement = [Replicate()]
+        query = DTensor.from_local(query, device_mesh, query_placement)
```

Comment:
could we use `distribute_tensor` instead for all inputs, so that we can have `query, key, value` as are global torch.Tensors, and `dist_query/key/value` that are DTensors, then we can run them separately and compare their final results to make sure numerics stay the same as a single device op.

you can look at an example here https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_matrix_ops.py#L276-L300

Reply:
that's functionally what this test does already just using .to_local()/full_tensor() instead

Reply:
done